### DISCUSSION TITLE
Section 5 - TSwap. Writing a test to prove swapExactInput will return 0

### DISCUSSION THREAD

**Discussion by JJScar** [Cyfrin/security-and-auditing-full-course-s23#206] [General]
Hello all!

I am trying to write the test to prove that the swapExactInput function will return zero as they did not use the return variable in the function. This is what i have so far, however I cannot get my head around what i am doing wrong:

```javascript
function testSwapExactInputAlwaysReturnsZero() public {
        vm.startPrank(liquidityProvider);
        weth.approve(address(pool), type(uint256).max);
        poolToken.approve(address(pool), type(uint256).max);
        pool.deposit(STARTING_Y, STARTING_Y, STARTING_Y, uint64(block.timestamp));
        vm.stopPrank();

        uint256 result = pool.swapExactInput(weth, STARTING_X, poolToken, 1e18, uint64(block.timestamp));
        assert(result == 0);
    }
```

**Comment by EngrPips** [comment]
Hello @JJScar, So below is the function you are talking about, and it specifies that it `returns(uint256 output)`. Considering that we are using solidity 0.8.20, I believe implicit return is a thing in solidity version 0.8.20, so you don't need to explicitly return `output` in the function; instead, you need to assign to it a value, and solidity would automatically return whatever value is assigned to it at the end of all the computation.
```js
// SPDX-License-Identifier: GNU General Public License v3.0
pragma solidity 0.8.20;

....

function swapExactInput(
        IERC20 inputToken,
        uint256 inputAmount,
        IERC20 outputToken,
        uint256 minOutputAmount,
        uint64 deadline
    )
        public
        revertIfZero(inputAmount)
        revertIfDeadlinePassed(deadline)
        //@reported: you never returned this value [Low]
        returns (uint256 output)
    {
        uint256 inputReserves = inputToken.balanceOf(address(this));
        uint256 outputReserves = outputToken.balanceOf(address(this));

        uint256 outputAmount = getOutputAmountBasedOnInput(
            inputAmount,
            inputReserves,
            outputReserves
        );

        if (outputAmount < minOutputAmount) {
            revert TSwapPool__OutputTooLow(outputAmount, minOutputAmount);
        }

        _swap(inputToken, inputAmount, outputToken, outputAmount);
    }
```

**Reply by JJScar** [reply]
Thanks for the reply! Yea I understand what is needed to fix and how to fix it. I am just trying to practice on my testing. How would I go about testing this function to prove that It returns 0. Because the way I tried did not get me there....

**Comment by EngrPips** [comment]
> Hello all!
> 
> I am trying to write the test to prove that the swapExactInput function will return zero as they did not use the return variable in the function. This is what i have so far, however I cannot get my head around what i am doing wrong:
> 
> ```js
> function testSwapExactInputAlwaysReturnsZero() public {
>         vm.startPrank(liquidityProvider);
>         weth.approve(address(pool), type(uint256).max);
>         poolToken.approve(address(pool), type(uint256).max);
>         pool.deposit(STARTING_Y, STARTING_Y, STARTING_Y, uint64(block.timestamp));
>         vm.stopPrank();
> 
>         uint256 result = pool.swapExactInput(weth, STARTING_X, poolToken, 1e18, uint64(block.timestamp));
>         assert(result == 0);
>     }
> ```

You can `console.log` the result to see that it is not `0.`

**Comment by JJScar** [comment]
For anyone who stumbled upon the same issue, it took my some time, but I got there (lol). Here is the test the passed:

```javascript
function testSwapExactInputAlwaysReturnsZero() public {
        vm.startPrank(liquidityProvider);
        weth.approve(address(pool), type(uint256).max);
        poolToken.approve(address(pool), type(uint256).max);
        pool.deposit(STARTING_Y, STARTING_Y, STARTING_Y, uint64(block.timestamp));
        vm.stopPrank();

        uint256 newAmount = 1e18;
        poolToken.mint(user, newAmount);
        vm.startPrank(user);
        poolToken.approve(address(pool), type(uint256).max);
        uint256 returned = pool.swapExactInput(poolToken, newAmount, weth, 0, uint64(block.timestamp));
        vm.stopPrank();

        assertEq(returned, 0);
    }
```

**Reply by EngrPips** [reply]
Hmmm, it's Interesting that you figured out a way to achieve what you wanted to achieve.
